src/runner.rs

```diff
@@ -172,43 +167,228 @@ impl Requirements {
 }
 
-impl private::Sealed for Path {}
-
-impl TestFnPath for Path {
-    fn convert(f: fn(&Self) -> Result<()>) -> TestFn {
-        TestFn::Path(f)
+impl TestFnLoadString {
+    fn call(&self, path: &Utf8Path) -> Result<()> {
+        let contents = std::fs::read_to_string(path)
+            .map_err(|err| format!("error reading file '{path}' as UTF-8: {err}"))?;
+        match self {
+            TestFnLoadString::Path(f) => f(path.as_ref(), contents),
+            TestFnLoadString::Utf8Path(f) => f(path, contents),
+        }
     }
 }
 
-impl private::Sealed for Utf8Path {}
-
-impl TestFnPath for Utf8Path {
-    fn convert(f: fn(&Self) -> Result<()>) -> TestFn {
-        TestFn::Utf8Path(f)
-    }
+#[derive(Clone, Copy)]
+#[doc(hidden)]
+pub enum TestFnLoadBinary {
+    Path(fn(&Path, Vec<u8>) -> Result<()>),
+    Utf8Path(fn(&Utf8Path, Vec<u8>) -> Result<()>),
 }
 
-impl TestFn {
+impl TestFnLoadBinary {
     fn call(&self, path: &Utf8Path) -> Result<()> {
+        let contents =
+            std::fs::read(path).map_err(|err| format!("error reading file '{path}': {err}"))?;
         match self {
-            TestFn::Path(f) => f(path.as_ref()),
-            TestFn::Utf8Path(f) => f(path),
+            TestFnLoadBinary::Path(f) => f(path.as_ref(), contents),
+            TestFnLoadBinary::Utf8Path(f) => f(path, contents),
+        }
+    }
+}
+
+/// We use a variant of autoref specialization to allow `TestFn` to be called with functions of one
+/// of several types. See
+/// <https://github.com/dtolnay/case-studies/blob/master/autoref-specialization/README.md>.
+///
+/// Notes:
+///
+/// We have to implement `PathKind` for all `F: Fn(&Path) -> Result<()>`, etc, because simply
+/// implementing it for `fn(&Path) -> Result<()>` won't perform a coercion from the concrete
+/// function type to the generic pointer. It means that technically you can call `kind()` on a
+/// non-free function, but it will be rejected by the `new` method.
+#[doc(hidden)]
+pub mod autoref {
+
+    use super::*;
+
+    mod private {
+        // We need to define a separate Sealed for each of the tags below, because Rust doesn't allow
+        // multiple kinds of F: Fn(T) -> Result<()> to implement the same trait.
+        pub trait PathSealed {}
+        pub trait Utf8PathSealed {}
+        pub trait PathStringSealed {}
+        pub trait Utf8PathStringSealed {}
+        pub trait PathBytesSealed {}
+        pub trait Utf8PathBytesSealed {}
+    }
+
+    // -- Paths --
+
+    #[doc(hidden)]
+    pub struct PathTag;
+
+    impl PathTag {
+        #[inline]
+        pub fn resolve(self, f: fn(&Path) -> Result<()>) -> TestFn {
+            TestFn::Path(f)
+        }
+    }
+
+    #[doc(hidden)]
+    pub trait PathKind: private::PathSealed {
+        #[inline]
+        fn kind(&self) -> PathTag {
+            PathTag
+        }
+    }
+
+    impl<F: Fn(&Path) -> Result<()>> private::PathSealed for F {}
+    impl<F: Fn(&Path) -> Result<()>> PathKind for F {}
```

Comment:
Decided to use copy-paste rather than a macro since I figured it wasn't that much code being duplicated.

Reply:
I couldn't figure out how to make this code generic over `&Path` and `&Utf8Path`, though, since we need to store concrete function pointer types for the polymorphic dispatch.

Reply:
Yeah doing this without a macro makes whats happening a bit easier to understand imo.

Reply:
Yeah given the Fn pointer can't be generic there isn't a clear way I can think of doing this (in the short time thinking about this problem), that and the required extra 'kind' and 'tag' impls isn't an issue

Reply:
Expanded the top-of-module documentation to talk about what's going on here.